docs/architecture/engine.md

```diff
@@ -0,0 +1,197 @@
+# Storage Engine Overview
+
+M3DB is a time series database that was primarily designed to be horizontally scalable and handle a large volume of monitoring time series data.
+
+## Caveats / Limitations
+
+1. M3DB currently supports exact ID based lookups. It does not support tag/secondary indexing. This feature is under development and future versions of M3DB will have support for a built-in reverse index.
+2. M3DB does not support updates / deletes. All data written to M3DB is immutable.
+3. M3DB does not support writing arbitrarily into the past and future. This is generally fine for monitoring workloads, but can be problematic for traditional [OLTP](https://en.wikipedia.org/wiki/Online_transaction_processing) and [OLAP](https://en.wikipedia.org/wiki/Online_analytical_processing) workloads. Future versions of M3DB will have better support for writes with arbitrary timestamps.
+4. M3DB does not support writing datapoints with values other than double-precision floats. Future versions of M3DB will have support for storing arbitrary values.
+5. M3DB does not support storing data with an indefinite retention period, every namespace in M3DB is required to have a retention policy which specifies how long data in that namespace will be retained for. While there is no upper bound on that value (Uber has production databases running with retention periods as high as 5 years), its still required and generally speaking M3DB is optimized for workloads with a well-defined [TTL](https://en.wikipedia.org/wiki/Time_to_live).
+6. Writing / reading from languages other than Golang. Communicating with M3DB currently requires use of a "fat" (logic-heavy) client, and there is only a Golang implementation at the moment.
```

Comment:
Not true per the last comment. We can use JSON over HTTP to hit the various thrift rpc endpoints. The `cluster` thrift service in each node wraps session APIs so we can still use it do the fanout to all nodes.

E.g. 

```
rungta@<redacted>:~$ cat request.json
{
  "nameSpace": "metrics",
  "rangeStart": 0,
  "rangeEnd": 1516989147,
  "id": "abc"
}

rungta@<redacted>:~$ curl -X POST  "http://0.0.0.0:6156/fetch" -d "@request.json"
{"datapoints":[]}
```

Reply:
do we want to document this? Could someone use this for a reasonable sized workload?

Reply:
Yes, but I'm not sure where exactly :| And ha, def not. We don't expose all the 'performant' methods on the `cluster` iirc